Replacing -d/--daemon with daemon command. Implementation of 'docker daemon' command

shishir-a412ed opened the pull request:
Signed-off-by: Shishir Mahajan <shishir.mahajan@redhat.com

aidanhs commented on docker/docker.go:
```diff
@@ -25,6 +26,14 @@ const (
+			os.Args[j] = "daemon"
+		}
+	}
+
```
This isn't backwards compatible. Consider `sudo docker -g ./graph -p ./d.pid -d`.

shishir-a412ed replied:
Can you elaborate ? All we are doing is replacing '-d/--daemon' with 'daemon' keyword. So that docker daemon is started using `sudo docker daemon` rather than `sudo docker -d`

shishir-a412ed commented:
@aidanhs 
We are trying to register `docker daemon` as a command. Like any other docker command, we should pass options after the command (daemon). So this should work `sudo docker daemon -g ./graph -p ./d.pid`

rhatdan commented:
@aidanhs The problem we have now, is users specifying commands that are only appropriate for the --daemon on other commands, and the tool silently allowing it.  Users expect something to happen, and have no idea why it did not.  -d needs to die.

aidanhs commented:
I absolutely agree that `-d` needs to die. I was planning on making a PR to warn users about bad arguments since this got stalled somewhat last time.

However, it's pretty easy to retain *full* backwards compat for a few releases - in the loop I've commented on, if you see a `-d` then remove it and insert `daemon` at position 1 in os.Args list. That would support the example I gave.

Alternatively, drop support for `-d` entirely. Please don't leave it the current state where it may or may not work (and possibly will fail in a cryptic way because the `-d` -> `daemon` transform is silent), depending on whether you had the foresight to give the `-d` arguments before anything else.

shishir-a412ed commented:
@aidanhs It works as long as we pass the arguments at the right place.
If you pass the argument before the `daemon` keyword. It takes it as a flag, and since it is not a flag (it fails), but an argument to the command daemon. it should be passed after the command, not before it.

We are not completely removing -d / --daemon to give users some head ups, as this command is deprecated and will be removed in near future. I have seen this before with other docker commands.

thaJeztah commented:
Jenkins failed on [`docker rmi .... docker rmi requires a minimum of 1 argument`](https://jenkins.dockerproject.com/job/Docker-PRs/254/consoleText). Don't think it's related to the PR

rhatdan commented:
@shishir-a412ed Change the patch to put the daemin to the front and @aidanhs will be happy and we will be backwards compatible.

rhatdan commented:
@thaJeztah Yes that is strange, possible jenkens is a little hosed up.

thaJeztah commented on docker/docker.go:
```diff
@@ -25,6 +26,14 @@ const (
+	for j := 1; j < len(os.Args); j++ {
+		if os.Args[j] == "-d" || os.Args[j] == "--daemon" {
+			log.Printf("%s is deprecated. Please use daemon command", os.Args[j])
+			os.Args[j] = "daemon"
```
to be backwards compatible, `daemon` should always become the *first* argument, even if `-d` was passed as the x-th argument. The other arguments should be re-ordered to come *after* `daemon`

aidanhs replied:
:+1:

aidanhs replied:
Actually, realisation that this suggestion isn't sufficient - see https://github.com/docker/docker/pull/10535#issuecomment-72765137 below.

i.e. you make it impossible to pass global-level arguments if you shuffle the daemon command to the front.

If #10453 was merged, then this PR was rebased on top of it, we'd have a decent way of filtering os.Args into two lists: global args and daemon args. Without that we're still treating api-enable-cors as a global arg when this PR gets merged, which is wrong.

thaJeztah commented on docker/flags.go:
```diff
@@ -38,7 +38,7 @@ func getDaemonConfDir() string {
 var (
 	flVersion     = flag.Bool([]string{"v", "-version"}, false, "Print version information and quit")
-	flDaemon      = flag.Bool([]string{"d", "-daemon"}, false, "Enable daemon mode")
+	//flDaemon      = flag.Bool([]string{"d", "-daemon"}, false, "Enable daemon mode")
```
Any reason to keep this in?

shishir-a412ed replied:
My bad ! I ll clean this up.

thaJeztah commented on daemon/server/daemon.go:
```diff
@@ -16,7 +17,7 @@ import (
 	"github.com/docker/docker/dockerversion"
 	"github.com/docker/docker/engine"
-	flag "github.com/docker/docker/pkg/mflag"
+	//flag "github.com/docker/docker/pkg/mflag"
```
Same here; Any reason to keep this in?

shishir-a412ed replied:
Same as above.

thaJeztah commented on contrib/init/systemd/docker.service:
```diff
@@ -5,8 +5,9 @@ After=network.target docker.socket
-ExecStart=/usr/bin/docker -d -H fd://
+ExecStart=/usr/bin/docker daemon -H fd://
 MountFlags=slave
+Restart=on-failure
```
Is adding `Restart=on-failure` here related to this PR?

shishir-a412ed replied:
@rhatdan Dan, this was added as part of your implementation (PR 9255). Should we remove this one ?
It was not part of the upstream.

thaJeztah replied:
Perhaps is something good, but if not directly related, it should be a separate PR

rhatdan replied:
I would remove it,  I have no idea where this came from.

shishir-a412ed replied:
@rhatdan removed this now.

thaJeztah commented on docs/sources/reference/api/docker_remote_api_v1.10.md:
```diff
@@ -1343,4 +1343,4 @@ stdout and stderr on the same socket. This might change in the future.
 "--api-enable-cors" when running docker in daemon mode.
 
-    $ docker -d -H="192.168.1.9:2375" --api-enable-cors
+    $ docker daemon -H="192.168.1.9:2375" --api-enable-cors
```
Not sure; should the older API docs also be updated? It's a bit of a difficult case; newer docker versions support the older API, whereas older docker versions *won't* support the new `daemon` command.

If the `-d` backwards compatibility stays in (for the time being), I *think* the older API docs should not be updated, or should mention both forms (`-d` / `daemon`).

Probably best to get some feedback from a doc maintainers for this.

jfrazelle commented:
I think you need a rebase it is failing on context errors https://jenkins.dockerproject.com/job/Docker-PRs/258/label=ubuntu-prs2/console

shishir-a412ed commented:
@jfrazelle I just rebased 2 mins ago.

jfrazelle commented:
you can see the logs that are failing in the link

thaJeztah commented:
I think what's missing in the PR is updates to the completion scripts.

According to this: https://github.com/docker/docker/pull/10416#issue-55782990, updating the [`fish` completions](https://github.com/docker/docker/blob/master/contrib/completion/fish/docker.fish) can be updated almost automatically.

The bash completions ([found here](https://github.com/docker/docker/tree/master/contrib/completion/bash)) need to be updated manually. You could ask @albers if he's willing to assist, but I think it's best to first see if the maintainers are willing to approve the PR first.

Not sure about the zsh completions, I think those have been neglected a bit.

albers commented:
If this is going to be accepted, I'd like to take over the bash completion part. @thaJeztah Thanks for pointing me to this.

tiborvass commented:
@shishir-a412ed Thanks for your contribution!

I also prefer having `docker daemon` instead of `docker -d`, however we cannot simply remove `docker -d` overnight. `docker -d` will have to live for a while. I'm okay to have a deprecation warning message for `docker -d`. Please don't make changes to your PR until we have decided, we wouldn't want to waste your time if we were to decide otherwise.

thaJeztah commented:
@tiborvass should this be labeled as "proposal"?

WRT backwards compatibility (also see the discussion above) I *think* that's almost in place, but help/usage probably changes, yes.

tiborvass commented:
@shishir-a412ed @thaJeztah Oh my bad, I tested with `-d -D` and saw:
```
INFO[0000] -d is deprecated. Please use daemon command
flag provided but not defined: -D
```

I assumed there was no backwards compat. Seems like it's *just* a bug in how the PR handles flags the old way. Also almost all the tests are failing, `docker run -d` is not working.

Anyway, I'll mark this as UX and it needs design approval.

shishir-a412ed commented:
@tiborvass I can take a look at those errors, and also work on backward compatibility issue.
However as you suggested, the first step is for you guys to make a decision on whether this is something docker wants as part of their product line ??

A quick decision on this would be really helpful as this is pending for a long time. PR 9255.

shishir-a412ed commented:
@tiborvass 

1) The reason why `docker -d -D` was not working, because -D (debug flag) was never part of Daemon Config struct. I have fixed this, you can test now.
2) I have fixed the `docker run -d` issue.
3) I have also fixed the backward compatibility issue. So now we can pass -d/--daemon at any position.

These were small changes, I don't mind fixing it. However we would really appreciate if we can get this change upstream. Let me know what you guys think ?

Shishir

aidanhs commented:
> -D (debug flag) was never part of Daemon Config struct

It still shouldn't be. It's perfectly valid to do `docker -D run blah` - it still needs to be a global option.
The point being made was same as the one I was making (that `docker -d -D` should transform into `docker -D daemon`) and is (theoretically) addressed by your point 3.

Unfortunately backwards compat (and normal functionality) is still broken. Some things which won't work (but should!) if this gets merged:
1. `docker -D run blah mycommand -a -b -c -d` (if we ignore the problem with `-D` being removed as a global arg)
2. `docker --tls -d`
3. `docker -d -H tcp://localhost:1234`
4. `docker tcp://localhost:1234 -H unix:///var/run/docker.sock daemon`

Each of these is broken for a different reason :)

shishir-a412ed commented:
The main problem seems to be the interleaving between global and daemon flags. The code is very tightly coupled !

@aidanhs for your point 3, I see the docker drone.io is also failing on that flag -H/--host.
If daemon is the only one using it, wouldn't it be best to move the HOST flag to daemon config struct ? Unlike `-D` (1) which you showed is more of a global argument (being used by both daemon and run)

aidanhs commented:
Daemon isn't the only one using `-H`, otherwise you'd never be able to connect to a modified daemon! E.g. `docker -H tcp://localhost:1234 run blah mycommand` connects to a docker daemon running in the appropriate place.

There isn't actually too much interleaving. Merging #10453 would give a reliable way of splitting args up because the daemon and global args would then be kept totally separate. If you make all flag creation go via the docker-created `opts` module, you can add some code to keep track of which arguments have been registered globally and for the daemon. Then you can use this introspection to reliably split up os.Args into `daemonArgs` and `globalArgs`.

aidanhs commented:
Alternatively, you could leave in the code for `flDaemon` (with deprecation messages on that code path) then duplicate the flags to add them to the new `daemon` subcommand. New flags would only be added to the subcommand and in time `flDaemon` would be removed. This is probably the simplest approach, but still needs some care (and the PR I keep plugging).

rhatdan commented:
I don't want the daemon flags to be valid for the docker run command any longer.  That is the problem we are seeing with our customers.  There is confusion in the current tool over which options are global and which are just for the daemon.

If we can merge the two patches together and then move the global options to the front of the command if they are messed up and a -d is present, I would prefer that.

shishir-a412ed commented:
I agree with @rhatdan, there should be a clear bifurcation of which flags applies to daemon, and which are global flags.

@aidanhs 
Also, I don't see how merging #10453  would solve this problem. Moving {SocketGroup, EnableCors} to daemon config is the change in that PR. How would that allow -D to be passed to both daemon and run (in the example above) ? If we consider daemon as a command {which means we have to move that code to api/client/commands.go} and somehow manage to give it access to global flags.

aidanhs commented:
> I don't want the daemon flags to be valid for the docker run command any longer. That is the problem we are seeing with our customers.

Yes, the only reason I'm spending time on this PR *because* I've been bitten by the global flags problem before. My patch (which was stage 1) moves some flags (which were incorrectly in the global config) into the daemon config. I was then going to follow up (stage 2) with a patch that would scan through any arguments before the subcommand and violently complain if 1) `-d` hadn't been specified and 2) you try to use a daemon flag. This is *only* possible with that PR in place! If it's not there you have you create your own blacklist of flags because the global/daemon separation is incorrect.

Converting `-d` to `daemon` would then have been stage 3. Because this PR is skipping stage 2 it needs to implement the code that would've been there anyway.

> If we can merge the two patches together and then move the global options to the front of the command if they are messed up and a -d is present, I would prefer that.

Yes, in order to have any hope of converting `-d` to `daemon` you need to have a *reliable* way of identifying daemon args and global args so you can split them up and put them in the right place. My PR is a prerequisite for that reliability, as otherwise you have daemon flags living in the global config.

To give an idea of how I would implement this in pseudocode if I had to do stage 2+3 together:

```
func main () {
    [...]
    dflagnames = []
    makeFlagHook(&dflagnames) // name of flags registered from now on will be inserted into dflagnames
    daemon.InstallFlags() // dflagnames now looks like ["--mtu", "--exec-driver", ...]
    flag.Parse()
    hasDFlag = False
    daemonFlags = []
    globalFlags = []
    for f in flag.iter() {
        if f.flag == "-d": hasDFlag = True
        if f.flag in dflagnames:
            daemonFlags.append(f.flag)
            if f.hasVal: daemonFlags.append(f.val)
        else:
            globalFlags.append(f.flag)
            if f.hasVal: globalFlags.append(f.val)
    }
    if hasDFlag: // converting -d into daemon
        subcommandArgs = ["daemon"] + daemonFlags
    elif not hasDFlag and len(daemonFlags) > 0: // not -d but have daemon args before subcommand
        WARN_USER()
    else: // not -d and no daemon args before subcommand
        pass
```

i.e. parse the flags before any subcommand, split them into a list of daemon args and a list of global args, check if daemon args are allowable (only if `-d` is in use).

However, there's a lot of silence from Docker Inc and I can see that both this and my PR now need rebasing. I'm not going to bother doing anything else until there are actually signs of interest.

shishir-a412ed commented:
@tiborvass @jfrazelle Can you please check why the drone.io is failing on this ? its showing again a `docker rmi` issue which doesn't seem to be related to this PR.

@aidanhs thanks for the tip on maintaining two lists {daemon_args, global_flags}. I didn't read your complete post and the pseudocode, but I believe the central idea should be same.
For me I felt it was more of a shuffling (arguments) problem to make it backward compatible. This is how I did it.

1) Scanned all the incoming arguments into two lists, daemon_args and global_args. {Used 2 pre-populated hash maps to classify} 
2) put all the global args and their corresponding values behind "daemon" keyword.
3) put all the daemon args after "daemon" keyword.
4) Now the command is reassembled and should be backwards compatible. Now docker will handle this like any other command (load, save, run etc).

I should be able to rebase your PR (moving {SocketGroup, EnableCors} into daemon config struct) with this approach with minimal effort.

PS: I unit tested this code, and it works fine. Although more testing is required to check for corner cases.

jfrazelle commented:
If you rebase it will remove drone entirely, we do not use it anymore. So
as long as jenkins passes you are golden

On Tue, Feb 10, 2015 at 10:58 AM, Shishir <notifications@github.com> wrote:

> @tiborvass <https://github.com/tiborvass> @jfrazelle
> <https://github.com/jfrazelle> Can you please check why the drone.io is
> failing on this ? its showing again a docker rmi issue which doesn't seem
> to be related to this PR.
>
> @aidanhs <https://github.com/aidanhs> thanks for the tip on maintaining
> two lists {daemon_args, global_flags}. I didn't read your complete post and
> the pseudocode, but I believe the central idea should be same.
> For me I felt it was more of a shuffling (arguments) problem to make it
> backward compatible. This is how I did it.
>
> 1) Scanned all the incoming arguments into two lists, daemon_args and
> global_args. {Used 2 pre-populated hash maps to classify}
> 2) put all the global args and their corresponding values behind "daemon"
> keyword.
> 3) put all the daemon args after "daemon" keyword.
> 4) Now the command is reassembled and should be backwards compatible. Now
> docker will handle this like any other command (load, save, run etc).
>
> I should be able to rebase your PR (moving {SocketGroup, EnableCors} into
> daemon config struct) with this approach with minimal effort.
>
> PS: I unit tested this code, and it works fine. Although more testing is
> required to check for corner cases.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/docker/docker/pull/10535#issuecomment-73759887>.
>

shishir-a412ed commented:
@jfrazelle thanks for that :) I ll rebase it now.

shishir-a412ed commented:
@jfrazelle I think it was the Jenkins that was failing on `docker rmi`. Sorry I misquoted it as drone.io.
Let me rebase first and see if it solves the problem.

Thanks again for your prompt response !

aidanhs commented:
@shishir-a412ed 

> but I believe the central idea should be same

Yes, your step description is exactly what I suggested. Unfortunately your detection for `hasDaemon` is still flawed. My concerns about the increasing complexity of this approach at backwards compatibility aside, here are a couple of examples of commands that I'm pretty sure (still) won't work:

1. `docker create imagename mycommand -a -b -c -d`
2. `docker -H tcp://localhost:1234 -H unix:///var/run/docker.sock daemon`

Both of these are relatively easy to fix, but this is from a quick look at code - I have a suspicion I can find a couple more.

thaJeztah commented:
> here are a couple of examples of commands that I'm pretty sure (still) won't work

The second example (I think) *shouldn't* work, given that the `daemon` command should come directly after `docker` (i.e. `docker daemon [opts]`)

aidanhs commented:
But `-H` is a global arg, per an example I gave of `docker run` further up this thread.

tiborvass commented:
*design review with myself*

Ok, there are two problems here, and I suggest we do two PRs:
* do not treat daemon flags as global flags (will make @rhatdan happy). So this means that `docker -d run busybox true` should not print a help message, and `docker -s foo run busybox true` should not succeed either. This would be a bugfix.
* deprecate `-d` and `--daemon` to promote a command called `daemon`, while keeping backwards compatibility. This should be much easier once the first PR gets merged.

I suggest @shishir-a412ed closes this PR for now, and someone (can be @shishir-a412ed again if he wants to) open a new PR to address the first issue, once that's merged @shishir-a412ed can reopen this PR.

Maintainers, voice your opinion (+1/-1, arguments, etc.): @crosbymichael @tianon @jfrazelle @unclejack @icecrime @LK4D4

jfrazelle commented:
Yes I agree with @tibor on the proposed actions. This would also make for smaller PRs and less back and forth considering item 1 is a "bug fix"

shishir-a412ed commented:
@tiborvass ... carrying up on the conversation we had on irc.

1) `docker -d run busy box true` : This bug we discussed on irc, and is a corner case which I would happily fix once we reach a design decision. {-d coming before run, and will be replaced by daemon keyword in this case}
2) `docker -s foo run busy box true` would fail. As `-s` is not a global flag but a daemon argument.

These changes are already there in this PR.

Now before we close this PR, we should atleast reach a decision whether you guys want this change or not. As this was the same case which happened with PR # 9255. After quite a bit of discussion and implementation provided by @rhatdan, the PR was closed with no design decision.

So lets make a decision and then I ll open another PR for the implementation.

jfrazelle commented:
I also think at this point since there has already been a lot of hard work put in we should find a compromise or carry the PR ourselves to make the changes so as not to impose more work on anyone else. Also if it comes down to that I would be more than happy to make the changes to get this in.

tiborvass commented:
So to reiterate: I am a big +1 on this, design-wise. Haven't looked at implementation. I would still prefer this to be 2 PRs, or best compromise: two commits.

@jfrazelle and I are already +1 for this.

tiborvass commented:
I just realized this is the 3rd PR implementing this, the 1st one dating back almost a year...

So let's make sure we are all okay on the design:
* new `daemon` command, displayed under `Commands` in `docker help` output
* `daemon` command has its own flags like `-g`, and `-s` etc.
* `-d`, `--daemon`, and all other daemon-specific flags work just like today (except bugfix, see below), but are deprecated (produce warning) and will not be displayed in `docker help`
* fix bug when using deprecated daemon-flags with a command: `docker -d run busybox true`, `docker -s foo run busybox true`.
* make sure there's test coverages for a bunch of cases, `docker -d -D`, `docker -d run -d`

If you're okay with all this, please +1 the design, and if you have concerns or questions voice them now.

jfrazelle commented:
+1 design

On Tuesday, February 10, 2015, Tibor Vass <notifications@github.com> wrote:

> I just realized this is the 3rd PR implementing this, the 1st one dating
> back almost a year...
>
> So let's make sure we are all okay on the design:
>
>    - new daemon command, displayed under Commands in docker help output
>    - daemon command has its own flags like -g, and -s etc.
>    - -d, --daemon, and all other daemon-specific flags work just like
>    today (except bugfix, see below), but are deprecated (produce warning) and
>    will not be displayed in docker help
>    - fix bug when using deprecated daemon-flags with a command: docker -d
>    run busybox true, docker -s foo run busybox true.
>    - make sure there's test coverages for a bunch of cases, docker -d -D, docker
>    -d run -d
>
> If you're okay with all this, please +1 the design, and if you have
> concerns or questions voice them now.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/docker/docker/pull/10535#issuecomment-73831711>.
>

bfirsh commented:
Design looks great to me.

duglin commented:
Haven't look at the details of the code, but I think having a `docker daemon` command (instead of `docker -d`) makes a lot of sense.  So +1 to what @tiborvass said in https://github.com/docker/docker/pull/10535#issuecomment-73831711

SvenDowideit commented:
+1 design!!

rhatdan commented:
Ok Shishir and I met today on simplifying his code to check if there is a command specified. 
If not he will call his substitution routine which will move global options and their values to the front of the command, replace -d and --daemon with daemon and then leave the rest of the options.


docker --selinux-enabled -d -D

Would become 

docker -D daemon --selinux-enabled

Hopefully we will have something to review by tomorrow.

thaJeztah commented:
Thats great! Thanks for the joined effort!

shishir-a412ed commented:
@tiborvass @jfrazelle 

Any updates on the design review ? I already see five +1's on the design including you and Jessie.

How many more do we need to start discussing the implementation?
Also, the Jenkins is GREEN now, In case you want to take a look at the code. I have fixed all the issues, keeping the backward compatibility.

Feel free to ask me any questions to clarify the reasoning behind the code.

Shishir

aidanhs commented:
Just to continue being a killjoy, I decided to actually build this branch (since I'm now finding the code difficult to reason about) to try out a couple of immediate suspicions.

1. `./bundles/1.5.0-dev/binary/docker daemon --help` -> no output.
2. `./bundles/1.5.0-dev/binary/docker -l info daemon` -> `docker: 'daemon' is not a docker command. See 'docker --help'.`

Of course, you could also hit the fragility near the root cause of point 2 if you had a host called `search` which you wanted to use in `--insecure-registry search`. The problem is that the heuristic used to do the shuffling must either a) piggyback on flag parsing, b) partially reimplement flag parsing or c) be imperfect. C is the current choice and will punish anyone using an insecure registry with a hostname the same as any of the `docker` subcommands.

shishir-a412ed commented:
@aidanhs 

1) I fixed another issue in the end, which might have caused this. Its a small miss, I have fixed this.
It should work now.

2) `docker info` command takes 0 arguments. I don't know why you are passing 'daemon' at the end of it.
Just running `docker info` works fine for me.

Shishir

aidanhs commented:
```
$ ./bundles/1.5.0-dev/binary/docker 
Usage: docker [OPTIONS] COMMAND [arg...]
[...]
  -l, --log-level=info            Set the logging level
[...]
```

It is valid to pass arguments separated by a space rather than an =. e.g. `docker -l=info daemon` is exactly the same as `docker -l info daemon`.

shishir-a412ed commented:
@aidanhs 

Sorry I misunderstood your point 2. I thought you were trying to run `docker info` command.
Thanks for pointing that out. I have fixed that. It was not an issue with shuffling of arguments.

My shuffling module handles both cases where values are separated with a space and '=', hence keeping backward compatibility.

Since the global flag `-l` had a value INFO which was matching with a docker command (INFO), it was misinterpreting it. I have fixed that now, you can test it.

Shishir

shishir-a412ed commented:
@aidanhs I have reverted the code so `docker -l info daemon` will fail for you now.
The reason is still the same {global flag matching with a command name: INFO}, although I have to reevaluate how and when to call the shuffling module to cover other scenarios. 

I ll update you once I have a stable branch.

Shishir

shishir-a412ed commented:
@aidanhs Please check now. Let me know if you still face any issues.

aidanhs commented:
1. I don't even have to run this one to know that it's going to break if run `docker create` where one of the arguments to the command is `-d`. If you must hardcode things, it needs to be treated the same way as `run` and `exec`.

2. I also suspect it breaks if we've named a container `daemon` and we try do then do anything at all with the container while using that name (commit, kill, pause). Or tagged an image to be called `daemon` and do image manipulation things etc.

3. Then there's the (now slightly tweaked from my previous comment) open question of whether we should really punish a user for choosing a bad hostname of `run` and then trying to do `docker --insecure-registry run daemon`. Though, thinking about it some more, `--insecure-registry` isn't a global arg...and neither is `--registry-mirror`. So if we move those then at least this problem goes away.

Can we revisit splitting this up into two PRs? Clearly this one's getting too long to remember as I've noted point 1 above before! https://github.com/docker/docker/pull/10535#issuecomment-73796611
I don't have time to continue being a human test suite (and I've not yet gone into actual code comments, this is just stuff that doesn't work).

shishir-a412ed commented:
For your point 1) I don't see any `-d` flag to `docker create` command when I run the `docker create` help menu. Which docker version are you using ?

I can take a look at point 2 and 3, although we are still in design discussion for now.

The code I have written is on my own accord. You are more than welcome to write your own better solution and provide one. The end goal for all of us is same: To introduce `docker daemon` command. 

No one is asking you to be a human test suite. your positive feedback is more than appreciated if you want to provide it. You are not obligated to do anything.

aidanhs commented:
As in, `docker create myimage mycommand -a -b -c -d`.

Ok, I guess I'll wait for design discussion to finish.

It is true, I don't need to contribute. I've put my time in because I think this PR is a worthy cause (same as you). I'm just noting that this PR is trying to do a lot which just makes it harder to get right. If there was a part 1 of "forbid daemon args where there's no daemon flag" it wouldn't need a design discussion (I assume) because it's a 'bugfix' rather than an alteration of well-defined reasonable-to-depend-upon behaviour.

A hypothetical: what if I were to put together a 'part 1' PR? If all that's going on here at the moment is a design discussion then in theory it could get merged with no impact on the discussion here? @shishir-a412ed, can you ping me your timezone on IRC and maybe we can figure something out?

shishir-a412ed commented:
Yes it's a tricky one ! If it were to be an isolated problem, things would be different.
But with all the build tags put in place, it has strict restrictions of what files can and cannot be included in different binaries.Then working towards a solution, keeping all the backward compatibility in mind, with lot of restrictions on moving the code around {Not to forget the go circular dependencies} .. makes it a much harder problem. So getting it right would obviously need inputs from everyone here.

If by point 1 you mean: Currently you can pass daemon args to any docker command {run, exec etc} as global flags, and docker doesn't complain. Yes that is a bug fix which needs to be dealt with.

This implementation deals with it, although if you want to take it in incremental steps then that would become a separate PR. Your point 2 and 3 are valid, and they are not hard to fix in the current implementation.

If you note the errors you are pointing out, are not related to shuffling module, but before or after when it is called. Shuffling module is actually in quite good shape. (I won't say the implementation is bug free right now, but can be brought to a stable stage with reasonable amount of testing)

The key contention here is (although I am jumping ahead to implementation discussion) whether the docker maintainers would be open to the idea of solving this problem using shuffling of args. If they are, rest can be fixed, as I have already made a lot of progress on this. 

If you still want to implement that bug fix, that's between you and docker upstream. if it gets merged, I would have to probably rebase my code on top of it. But keeping the complexity of the problem in mind, it would be better if we move together rather than going separate ways.

I am available 9 AM to 6 PM EST. N usually reply to e-mails, msgs until I am awake.

Shishir

shishir-a412ed commented:
@tiborvass any updates ?

ehazlett commented:
ping @tiborvass

calavera commented on docker/noDaemon.go:
```diff
@@ -0,0 +1,9 @@
+package main
+
+func installDaemonFlags(args ...string) error {
+
```
do you mind to remove the extra blank lines from this function?

calavera commented:
It looks like this PR needs to be rebased before merging.

shishir-a412ed commented:
@calavera this PR would require a lot more than rebasing :)
I had words with @tiborvass. We will be doing this in 2 parts.

Part 1 would fix an existing issue in docker.
https://github.com/docker/docker/pull/11418

Once that is merged. we ll discuss how to go about this one.

Shishir

icecrime commented:
*Collective review with @LK4D4 @calavera @jfrazelle @crosbymichael @tiborvass*

Go for `docker daemon` subcommand: pushing this to code review!

shishir-a412ed commented:
@icecrime
This PR probably needs rebasing, and some work before it can be reviewed. I ll update here in 1-2 days.

thaJeztah commented:
> Go for docker daemon subcommand: pushing this to code review!

\o/

shishir-a412ed commented:
@tiborvass with the implementation strategy we discussed on irc, I would need to use PR #11418 
to build the `docker daemon` command feature. 

Since it is currently unmerged {pending review}, the only way I see it is to use PR # 11418 code to build this PR as a working code. You guys can then review it and if you guys are okay with it. I ll remove the PR # 11418 code from this PR and we ll merge it as two separate PRs.

Let me know if this works for you, and I ll work on this PR.

Shishir

shishir-a412ed commented:
@tiborvass I have pushed the updated code for this PR. You guys can start reviewing the code.
I will fix the jenkins by that time. I will also add integration-tests while the code review progresses. 

Based on the feedback from you guys, I ll make the appropriate changes.
For help understanding the code, I have provided a detailed description below.

Shishir

shishir-a412ed commented:
PR: Replacing -d/--daemon with daemon command. Implementation of 'docker daemon' command
Description of the code changes starting from main()

1) docker/docker:go: installDaemonFlags () Calls:
        a) docker/daemon.go: newServerCli(): Create a server side object (srv) like client side object  (NewDockerCli)
        b) docker/daemon.go: srv.subcmd(): Creates an empty daemon flagset. This call will return the daemon flagset `cmd`
        c) docker/daemon.go: installDaemonFlags() will then set all the daemon flags into `cmd` flagset.

2) docker/docker.go: checkDaemon(): This function will determine if -d/--daemon is coming in os.Args and if shuffling is required to maintain backward compatibility.

NOTE: checkDaemon() uses 2 maps {flag.CommandLine.FlagsMap, cmd.FlagsMap} to achieve this.
More explanation on these maps in point (6)

3) If STEP 2 is TRUE, call docker/docker.go: shuffle(): This function will shuffle the args to maintain backward compatibility.
                                                        It will place all the global flags before -d/--daemon.
                                                        It will place all the daemon args after -d/--daemon
                                                        It will replace -d/--daemon with `daemon` keyword.
                                                        The output of this would be: `dockerbinary GLOBAL-FLAGS daemon DAEMON-ARGS`

NOTE: shuffle() uses 2 maps {flag.CommandLine.FlagsMap, cmd.FlagsMap} to achieve this.
More explanation on these maps in point (6)

4) If the user starts the daemon with 'docker daemon' command (which is recommended) no shuffling would be done. In this case, if the user tries to pass daemon args as global flags docker will blow up. This is expected behaviour.

5) Now we check flag.Args()[0] == "daemon" i.e. if it's a daemon command we set `*flDaemon=true`.

6) if *flDaemon is TRUE, we parse the daemon flags by calling `parseDaemonFlags()` and starts the daemon by calling `mainDaemon()`.

7) pkg/mflag/flag.go: `FlagsMap` is the map used to classify whether the incoming `os.Arg` is a daemon arg or global flag. And whether it is a boolean flag {value optional} OR non boolean flag {value mandatory}. `FlagsMap` is part of the `FlagSet struct` hence two maps would be populated, one for daemon flagset and other for global (CommandLine) flagset.

These 2 maps are populated when the flags are set in the `FlagSet`. Once populated these 2 maps can be used by checkDaemon() and shuffle() functions.

Shishir

aidanhs commented on docker/docker.go:
```diff
@@ -24,11 +24,118 @@ const (
+				if args[j] == command[0] {
+					return false
+				}
+			}
```
Why not just return false here instead of this loop?
If it's not a command it's probably a misspelling of a command, so there's no point in shuffling.

shishir-a412ed replied:
We need the loop to check if we have hit a command, so that we can return and go no further.
to make sure `-d` is not an argument to a command and is a daemon start (-d/--daemon) which requires shuffling.

aidanhs replied:
Ok, let me explain by giving a concrete example: `docker rnu myimage cmd -a -b -c -d`.
Clearly this is a misspelling of the `run` command. Unfortunately, because you only check for real subcommands, this will be (incorrectly) identified as a command to shuffle into a daemon command.
My proposal is: as soon as you find any argument which isn't a daemon or global argument we should assume that this is *not* a daemon command. This correctly handles a) valid subcommands b) misspelled subcommands.

rhatdan replied:
Seems reasonable to me.

aidanhs commented on docker/docker.go:
```diff
@@ -24,11 +24,118 @@ const (
+	//Shuffling of arguments required.Create global and rest args list.
+
+	for j := 1; j < len(args); j++ {
+		if args[j] != "-d" && args[j] != "--daemon" {
```
I don't know whether there's a docker style guide, but might be nice to write this as
```
if args[j] == "-d" || args[j] == "--daemon" {
    continue
}
```
which lets you unindent the rest of this block by one tab.

shishir-a412ed replied:
This is just a styling comment. Both are valid, so I ll wait for the maintainers to comment.

rhatdan replied:
I agree @aidanhs this would make it easier to understand.

aidanhs commented on docker/docker.go:
```diff
@@ -24,11 +24,118 @@ const (
+
+	for j := 1; j < len(args); j++ {
+		if args[j] != "-d" && args[j] != "--daemon" {
+			if args[j][:1] == "-" {
```
Again here, there's no `else` so I might prefer the `!= ... break` style and unindenting the rest. Might want to wait for a maintainer code review though.

aidanhs replied:
In fact, I'm not sure why there isn't an `else`. Admittedly, the only reason we'd hit it here is a bug in shuffle or checkDaemon...but I'm pretty sure dropping values on the floor isn't the behaviour we want in that situation.

aidanhs commented on docker/docker.go:
```diff
@@ -24,11 +24,118 @@ const (
+						//Daemon flag with a value
+						rest_args = append(rest_args, args[j])
+						continue
+					}
```
Do we end up dropping arguments we don't know about? Surely there should be a fall-through case?
In fact, I'd personally think it makes sense for the fall-through case to append it to daemon args so I think you could get rid of the check for whether the flag exists in `cmd.FlagsMap` and just unconditionally append it to `rest_args`.
(not a maintainer, wait for official comment etc)

shishir-a412ed replied:
We would need both maps, as those maps are populating which flags are boolean and which are not, thereby identifying which can take an optional argument. One way is to hard code all the flags in one single map. In order to avoid that we are populating maps as part of flagset. Since we have segregated the flagsets {daemon, global} two maps are dynamically generated for them.

aidanhs replied:
Well, for this particular case we know it's taking the optional argument as a single argument string (we've already seen the `=`).
Anyway, let's ignore this for now as it's a distraction from the main issue - dropping values on the floor. Suggested fall-through (same as below): `rest_args = append(rest_args, args[j])`.

aidanhs commented on docker/docker.go:
```diff
@@ -24,11 +24,118 @@ const (
+						//Daemon bool flag
+						rest_args = append(rest_args, args[j])
+						continue
+					}
```
Also dropping unknown args here.

shishir-a412ed replied:
This is a valid concern, I ll take a look on this.

shishir-a412ed commented:
@aidanhs @rhatdan I have made the requested changes.

shishir-a412ed commented:
@rhatdan 
adding some testing code to validate the behaviour now.

aidanhs commented on docker/docker.go:
```diff
@@ -24,11 +24,113 @@ const (
+		} else if args[j][:1] == "-" {
+			if index_of_equal := strings.Index(args[j], "="); index_of_equal != -1 {
+				continue //Flag with a value
+			} else {
```
Minor style note that you don't need the `else`, you can just remove it and unindent the block.

shishir-a412ed replied:
fixed.

aidanhs commented on docker/docker.go:
```diff
@@ -24,11 +24,113 @@ const (
+		if args[j] == "-d" || args[j] == "--daemon" {
+			return true
+		} else if args[j][:1] == "-" {
+			if index_of_equal := strings.Index(args[j], "="); index_of_equal != -1 {
```
Why not use string.Contains?

shishir-a412ed replied:
changed to `strings.Contains`.

aidanhs commented on docker/docker.go:
```diff
@@ -24,11 +24,113 @@ const (
+				//Daemon bool flag
+				rest_args = append(rest_args, args[j])
+				continue
+			}
```
You can get rid of this `else if` block entirely.

shishir-a412ed replied:
Fixed.

aidanhs commented on docker/docker.go:
```diff
@@ -37,6 +139,10 @@ func main() {
 	flag.Parse()
 	// FIXME: validate daemon flags here
```
Remove this fixme? You are fixing it!

shishir-a412ed replied:
Fixed.

aidanhs commented:
Will look closer tomorrow, but looks generally good to me.

Commit:
Replacing -d/--daemon with daemon command. Implementation of 'docker daemon' command

Signed-off-by: Shishir Mahajan <shishir.mahajan@redhat.com>

shishir-a412ed commented:
@tiborvass any updates on this? I have added the test cases as well.

tiborvass commented:
Thanks @shishir-a412ed I'll carry this PR and #11418 to get it in asap.

aidanhs commented:
@tiborvass can you clarify? Are you going to make this a PR as part of something else? Is there anywhere to track progress?

Just not quite sure what's happened to this PR...

icecrime commented:
@aidanhs What is means is that we want the feature, but not in its current form: @tiborvass will take on @shishir-a412ed and make the required changes to get it merged.

shishir-a412ed commented:
@icecrime Can you please elaborate a little bit ?
Would you be using this PR's code as part of another PR ?
Are you going to use the same strategy (shuffling for maintaining the backward compatibility, segregating daemon and global flags into two different flagsets {I assume segregation is something we all want} ) ? or you plan on solving this with a completely different approach ?

aidanhs commented:
I'm mainly interested in how I can follow progress, given @tiborvass probably has quite a few other things on his plate as well.

rhatdan commented:
@tiborvass @crosbymichael @jfrazelle Has this patch been moved somewhere in a new pull request that we can view?  Usually when you guys do this you refer to a new pull request where we can comment and continue to contribute.

We have been working on this pull request for over a year and just having it dissapear is not supporting the community.

duglin commented:
I think @tiborvass is working on a new PR but its not been submitted yet.

thaJeztah commented:
@duglin I agree it's a bit confusing though to "carry" without opening a new PR; it makes people nervous. Perhaps just opening a placeholder "WIP" PR would have kept people at ease :)

aidanhs commented:
Any update? Two week silence?

shishir-a412ed commented:
@tiborvass 
Any updates on this? It's been a month since you closed this and #11418.
I have put in a lot of effort on this one, and it is a little demotivating and demoralizing to see all my effort go in vain. 
how can I be sure that the next PR I spend my time and effort on would not become like this one?

/cc @crosbymichael @rhatdan

rhatdan commented:
I agree, this is VERY contributor unfriendly.  To say you are taking over a patch set and then going radio silent...

tiborvass commented:
I sincerely apologize to @shishir-a412ed and everyone on this thread, this was 100% my fault. I'll make sure not to do this again.

Here's the PR #13771 I just opened, it was dependent on #13636 that just got merged. I welcome your feedback there on #13771. Thanks
